packages/multichain-network-controller/src/MultichainNetworkController.ts

```diff
@@ -0,0 +1,204 @@
+} from './constants';
+import {
+  MULTICHAIN_NETWORK_CONTROLLER_NAME,
+  type MultichainNetworkControllerState,
+  type MultichainNetworkControllerMessenger,
+  type SupportedCaipChainId,
+} from './types';
+import {
+  checkIfSupportedCaipChainId,
+  getChainIdForNonEvmAddress,
+} from './utils';
+
+/**
+ * The MultichainNetworkController is responsible for fetching and caching account
+ * balances.
+ */
+export class MultichainNetworkController extends BaseController<
+  typeof MULTICHAIN_NETWORK_CONTROLLER_NAME,
+  MultichainNetworkControllerState,
+  MultichainNetworkControllerMessenger
+> {
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: MultichainNetworkControllerMessenger;
+    state?: Omit<
+      Partial<MultichainNetworkControllerState>,
+      'multichainNetworkConfigurationsByChainId'
+    >;
+  }) {
+    super({
+      messenger,
+      name: MULTICHAIN_NETWORK_CONTROLLER_NAME,
+      metadata: MULTICHAIN_NETWORK_CONTROLLER_METADATA,
+      state: {
+        ...getDefaultMultichainNetworkControllerState(),
+        ...state,
+      },
+    });
+
+    this.#subscribeToMessageEvents();
+    this.#registerMessageHandlers();
+  }
+
+  /**
+   * Sets the active EVM network.
+   *
+   * @param id - The client ID of the EVM network to set active.
+   */
+  async #setActiveEvmNetwork(id: NetworkClientId): Promise<void> {
+    // Notify listeners that setActiveNetwork was called
+    this.messagingSystem.publish(
+      'MultichainNetworkController:networkDidChange',
+      id,
+    );
+
+    // Indicate that the non-EVM network is not selected
+    this.update((state) => {
+      state.isEvmSelected = true;
+    });
+
+    // Prevent setting same network
+    const { selectedNetworkClientId } = this.messagingSystem.call(
+      'NetworkController:getState',
+    );
+
+    if (id === selectedNetworkClientId) {
+      // EVM network is already selected, no need to update NetworkController
+      return;
+    }
+
+    // Update evm active network
+    await this.messagingSystem.call('NetworkController:setActiveNetwork', id);
+  }
+
+  /**
+   * Sets the active non-EVM network.
+   *
+   * @param id - The chain ID of the non-EVM network to set active.
+   */
+  #setActiveNonEvmNetwork(id: SupportedCaipChainId): void {
+    if (id === this.state.selectedMultichainNetworkChainId) {
+      if (!this.state.isEvmSelected) {
+        // Same non-EVM network is already selected, no need to update
+        return;
+      }
+
+      // Indicate that the non-EVM network is selected
+      this.update((state) => {
+        state.isEvmSelected = false;
+      });
+
+      // Notify listeners that setActiveNetwork was called
+      this.messagingSystem.publish(
+        'MultichainNetworkController:networkDidChange',
+        id,
+      );
+    }
+
+    // Notify listeners that setActiveNetwork was called
+    this.messagingSystem.publish(
+      'MultichainNetworkController:networkDidChange',
+      id,
+    );
```

Comment:
Why do we notify twice? We just do it in the `if` right before.

Reply:
Good catch. It was missing a return statement. Ended up simplifying this logic